### PR TITLE
Sanitise asset name

### PR DESCRIPTION
**Pull request by ffranr**
Closes https://github.com/lightninglabs/taproot-assets/issues/506
Closes https://github.com/lightninglabs/taproot-assets/issues/523

### CODE REVIEW

File: asset/asset.go [hunk 1]
```diff
@@ -18,6 +18,13 @@ import (
 	"github.com/lightningnetwork/lnd/tlv"
 )
 
+const (
+	// MaxAssetNameLength is the maximum byte length of an asset's name.
+	// This byte length is equivalent to character count for single-byte
+	// UTF-8 characters
```
Comment (guggero) [inline asset/asset.go:24]: nit: missing full stop at end of sentence.

File: rpcserver.go [hunk 2]
```diff
@@ -302,9 +302,9 @@ func (r *rpcServer) MintAsset(ctx context.Context,
 		return nil, fmt.Errorf("asset cannot be nil")
 	}
 
-	// An asset name is mandatory, and cannot be the empty string.
-	if len(req.Asset.Name) == 0 {
-		return nil, fmt.Errorf("asset name cannot be empty")
+	err := asset.ValidateAssetName(req.Asset.Name)
```
Comment (guggero) [inline rpcserver.go:305]: Should we also invoke this function in `Seedling.validateFields()`?

**Review by guggero — APPROVED** [review]
Very nice, LGTM :tada:

**Comment by jharveyb** [comment]
I think we could make progress here without clamping to ASCII - something like `strings.ToValidUTF8()` + `unicode.IsPrint()` would filter out control characters IIRC without filtering out other languages.

https://pkg.go.dev/strings#ToValidUTF8
https://pkg.go.dev/unicode#IsPrint

**Comment by ffranr** [comment]
@jharveyb , good idea.

I've made use of `unicode.IsPrint()`. I think that should take account of valid UTF8 (so I don't think we need `strings.ToValidUTF8()`). Let me know what you think!

**Comment by jharveyb** [comment]
> @jharveyb , good idea.
> 
> I've made use of `unicode.IsPrint()`. I think that should take account of valid UTF8 (so I don't think we need `strings.ToValidUTF8()`). Let me know what you think!

I think having invalid UTF-8 in the string when iterating over it would cause an issue and possibly incorrect decoding. I think we could use `IsValidUTF8()` to detect if the entire string is valid before we actually handle each rune (and exit early if it's invalid).

Re: control characters, not sure if we should escape them or reject them. They are not `IsPrint()` IIUC.

File: asset/asset.go [hunk 3]
```diff
@@ -1045,3 +1053,64 @@ func (a *Asset) Leaf() (*mssmt.LeafNode, error) {
...
+	}
+	return nil
+}
+
+// escapeWhiteSpace escapes the given character if it is a white space
+// character.
+func escapeWhiteSpace(char int32) string {
+	switch char {
+	case ' ':
+		return "\\s"
```
Comment (guggero) [inline asset/asset.go:1074]: I think white spaces are probably okay in the name?

File: asset/asset.go [hunk 4]
```diff
@@ -1045,3 +1053,64 @@ func (a *Asset) Leaf() (*mssmt.LeafNode, error) {
...
+	if len(name) > MaxAssetNameLength {
+		return fmt.Errorf("asset name cannot exceed %d bytes",
+			MaxAssetNameLength)
+	}
+	return nil
+}
+
+// escapeWhiteSpace escapes the given character if it is a white space
+// character.
+func escapeWhiteSpace(char int32) string {
```
Comment (guggero) [inline asset/asset.go:1071]: Should we use the internal `rune` type to make this a bit more explicit?

**Review by guggero — APPROVED** [review]
LGTM :tada:

File: asset/asset.go (hunk 4 above)
Comment (ffranr) [reply asset/asset.go:1071]: Good point!

**Comment by jharveyb** [comment]
Looks good! I think `escapeWhitespace()` is no longer needed as those chars are already rejected by the `IsPrint()` check.

Would be good to have some test cases for this; can pull different string types from https://go.dev/blog/strings

File: rpcserver.go (hunk 2 above)
Comment (Roasbeef) [inline rpcserver.go:305]: Do we also need to mirror this when validating?

File: asset/asset.go [hunk 5]
```diff
@@ -1051,5 +1061,31 @@ func ValidateAssetName(name string) error {
 	if len(name) == 0 {
 		return fmt.Errorf("asset name cannot be empty")
 	}
+
+	// Ensure the asset name is not too long.
+	if len(name) > MaxAssetNameLength {
```
Comment (Roasbeef) [inline asset/asset.go:1066]: Confirmed these checks still allow emojis ;)

File: asset/asset.go [hunk 6]
```diff
@@ -18,6 +21,13 @@ import (
 	"github.com/lightningnetwork/lnd/tlv"
 )
 
+const (
+	// MaxAssetNameLength is the maximum byte length of an asset's name.
+	// This byte length is equivalent to character count for single-byte
+	// UTF-8 characters.
+	MaxAssetNameLength = 64
```
Comment (Roasbeef) [inline asset/asset.go:28]: If we want to enforce this as a protocol level check, then we also need to apply this during proof verification.

**Review by Roasbeef — CHANGES_REQUESTED** [review]

File: asset/asset.go (hunk 6 above)
Comment (ffranr) [reply asset/asset.go:28]: Added a commit to address this.

**Comment by ffranr** [comment]
I've added an extra commit which will validate the asset's name during proof verification.
Let me know what you guys think @jharveyb @Roasbeef

**Review by guggero — APPROVED** [review]
Looks great, LGTM :tada:

**Review by jharveyb — APPROVED** [review]
Looks good! 🚀

**Pull request merged**